src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs

```diff
@@ -2,294 +2,450 @@
-        }
-        public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, out string? error)
-        {
-            error = null;
-
-            // validate type before calculating intrinsic gas to avoid exception
-            return ValidateTxType(transaction, releaseSpec, ref error)
-                   // This is unnecessarily calculated twice - at validation and execution times.
-                   && ValidateWithError(transaction.GasLimit >= IntrinsicGasCalculator.Calculate(transaction, releaseSpec), TxErrorMessages.IntrinsicGasTooLow, ref error)
-                   // if it is a call or a transfer then we require the 'To' field to have a value while for an init it will be empty
-                   && ValidateWithError(ValidateSignature(transaction, releaseSpec), TxErrorMessages.InvalidTxSignature, ref error)
-                   && ValidateChainId(transaction, ref error)
-                   && ValidateWithError(Validate1559GasFields(transaction, releaseSpec), TxErrorMessages.InvalidMaxPriorityFeePerGas, ref error)
-                   && ValidateWithError(Validate3860Rules(transaction, releaseSpec), TxErrorMessages.ContractSizeTooBig, ref error)
-                   && Validate4844Fields(transaction, ref error);
+            error = TxErrorMessages.InvalidTxType(releaseSpec.Name);
+            return false;
         }
 
-        private static bool Validate3860Rules(Transaction transaction, IReleaseSpec releaseSpec) =>
-            !transaction.IsAboveInitCode(releaseSpec);
+        return validator.IsWellFormed(transaction, releaseSpec, out error);
+    }
+}
+
+public sealed class AllTxValidator(List<ITxValidator> validators) : ITxValidator
+{
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
 
-        private static bool ValidateTxType(Transaction transaction, IReleaseSpec releaseSpec, ref string error)
+        foreach (ITxValidator validator in validators)
         {
-            bool result = transaction.Type switch
+            if (!validator.IsWellFormed(transaction, releaseSpec, out error))
             {
-                TxType.Legacy => true,
-                TxType.AccessList => releaseSpec.UseTxAccessLists,
-                TxType.EIP1559 => releaseSpec.IsEip1559Enabled,
-                TxType.Blob => releaseSpec.IsEip4844Enabled,
-                _ => false
-            };
-
-            if (!result)
-            {
-                error = TxErrorMessages.InvalidTxType(releaseSpec.Name);
                 return false;
             }
+        }
 
+        return true;
+    }
+}
+
+public sealed class IntrinsicGasTxValidator : ITxValidator
+{
+    public static readonly IntrinsicGasTxValidator Instance = new();
+    private IntrinsicGasTxValidator() { }
+
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
```

Comment:
if we made error a `ref` we could simplify the code a bit  - as it was originally

Reply:
I don't see how we could apply this change without changing the `ITxValidator` interface.
Could you elaborate a bit on how the code would look?

Reply:
We ended up going with a `ValidatorResult` return value rather than the `TryX` approach of returning a `bool` with an associated `out` parameter.